falcon/util/streams.py

```diff
@@ -0,0 +1,276 @@
+import functools
+import io
+
+DEFAULT_CHUNK_SIZE = io.DEFAULT_BUFFER_SIZE * 4
+
+
+class BufferedStream:
+
+    def __init__(self, read, max_stream_len, chunk_size=None):
+        self._read_func = read
+        self._chunk_size = chunk_size or DEFAULT_CHUNK_SIZE
+
+        self._buffer = b''
+        self._buffer_len = 0
+        self._buffer_pos = 0
+        self._max_bytes_remaining = max_stream_len
+
+    def peek(self, amount=-1):
+        # PERF(vytas) In Cython, bind types:
+        #   cdef Py_ssize_t read_amount
+
+        amount = int(amount)
+        if amount < 0 or amount > self._chunk_size:
+            amount = self._chunk_size
+
+        if self._buffer_len - self._buffer_pos < amount:
+            read_amount = (self._chunk_size - self._buffer_len +
+                           self._buffer_pos)
+            if read_amount >= self._max_bytes_remaining:
+                read_amount = self._max_bytes_remaining
+
+            if self._buffer_pos == 0:
+                self._buffer += self._read_func(read_amount)
+                self._buffer_len = len(self._buffer)
+            else:
+                self._buffer = (self._buffer[self._buffer_pos:] +
+                                self._read_func(read_amount))
+                self._buffer_len = len(self._buffer)
+                self._buffer_pos = 0
+            self._max_bytes_remaining -= read_amount
+
+        return self._buffer[self._buffer_pos:self._buffer_pos + amount]
+
+    def _normalize_size(self, amount):
+        # PERF(vytas) In Cython, bind types:
+        #   cdef Py_ssize_t result
+        #   cdef Py_ssize_t max_amount
+
+        max_amount = (self._max_bytes_remaining + self._buffer_len -
+                      self._buffer_pos)
+
+        if amount is None or amount == -1 or amount > max_amount:
+            return max_amount
+        return amount
+
+    def read(self, amount=-1):
+        return self._read(self._normalize_size(amount))
+
+    def _read(self, amount):
+        # PERF(vytas) In Cython, bind types:
+        #   cdef Py_ssize_t read_amount
+        if self._buffer_len == 0:
+            self._max_bytes_remaining -= amount
+            return self._read_func(amount)
+
+        if amount == self._buffer_len and self._buffer_pos == 0:
+            result = self._buffer
+            self._buffer_len = 0
+            self._buffer = b''
+            return result
+
+        if amount < self._buffer_len - self._buffer_pos:
+            self._buffer_pos += amount
+            return self._buffer[self._buffer_pos - amount:self._buffer_pos]
+
+        # NOTE(vytas): if amount > self._buffer_len - self._buffer_pos
+        read_amount = amount - self._buffer_len + self._buffer_pos
+        result = self._buffer[self._buffer_pos:]
+        self._buffer_len = 0
+        self._buffer_pos = 0
+        self._buffer = b''
+        self._max_bytes_remaining -= read_amount
```

Comment:
See also my other comment re trusting `self._read_func()` to return the requested number of bytes (it could be less).

Reply:
This approach is ported from `BoundedStream`:
```python
        # NOTE(kgriffs): Default to reading all remaining bytes if the
        # size is not specified or is out of bounds. This behaves
        # similarly to the IO streams passed in by non-wsgiref servers.
        if (size is None or size == -1 or size > self._bytes_remaining):
            size = self._bytes_remaining

        self._bytes_remaining -= size
        return target(size)
```

Reply:
OK, seems like we need to fix it over there as well then.